### PR TITLE
Make lms more resilient to marketing being down

### PR DESCRIPTION
**Pull request by shadinaif**
Make lms more resilient to marketing being down

* Set 3 seconds timeout for get_marketing_data

-------------
Related to https://edraak.atlassian.net/browse/OU-399

### CODE REVIEW

File: lms/djangoapps/course_api/helpers.py [hunk 1]
```diff
@@ -53,9 +53,14 @@ def get_marketing_data(course_key, language):
...
     url = marketing_root_format.format(course_id=course_key)
 
-    response = requests.get(url=url, headers={
-        'Accept-Language': language,
-    })
+    try:
+        response = requests.get(
+            url=url,
+            headers={'Accept-Language': language},
+            timeout=3
```
Comment (ghost) [inline lms/djangoapps/course_api/helpers.py]: Make it a configurable setting please.

```suggestion
            timeout=settings.MARKETING_API_TIMEOUT,
```

File: lms/djangoapps/course_api/helpers.py [hunk 2]
```diff
@@ -53,9 +53,14 @@ def get_marketing_data(course_key, language):
...
+        )
+    except requests.exceptions.Timeout:
+        response = HttpResponseNotFound()
```
Comment (ghost) [inline lms/djangoapps/course_api/helpers.py]:
```suggestion
from rest_framework import status   # On the top of the file.
# .......
        raise APIException(
                {"status_code": status.HTTP_503_SERVICE_UNAVAILABLE, "developerMessage": "Marketing courses API have timed out."},
                code=status.HTTP_503_SERVICE_UNAVAILABLE,
            ) 
```

Comment (ghost) [inline lms/djangoapps/course_api/helpers.py]: Also add another check for all other `Exception` to make the API return more relevant information. I'd try it first myself to see the difference between the two behaviours.

```suggestion
        except Exception:
            log.exception('Something went wrong with the marketing courses API')
            raise APIException(
                {"status_code": status.HTTP_500_INTERNAL_SERVER_ERROR, "developerMessage": "Marketing courses didn't respond correctly."},
                code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            ) 

```

**Review by ghost — CHANGES_REQUESTED** [review]
Thanks @shadinaif. I've added a couple of comments before we can proceed with this PR.

File: lms/djangoapps/course_api/helpers.py (hunk 1 above)
Comment (shadinaif) [reply lms/djangoapps/course_api/helpers.py]: Done

File: lms/djangoapps/course_api/helpers.py (hunk 2 above)
Comment (shadinaif) [reply lms/djangoapps/course_api/helpers.py]: Done

Comment (shadinaif) [reply lms/djangoapps/course_api/helpers.py]: Done

File: cms/envs/aws.py [hunk 3]
```diff
@@ -523,6 +523,9 @@
 
 PROGS_URLS = ENV_TOKENS.get('PROGS_URLS', PROGS_URLS)
 
+# API Calls Timeout
+EDRAAK_MARKETING_API_TIMEOUT = ENV_TOKENS.get("EDRAAK_MARKETING_API_TIMEOUT", 3)
```
Comment (ghost) [inline cms/envs/aws.py]:
```suggestion
EDRAAK_MARKETING_API_TIMEOUT = ENV_TOKENS.get('EDRAAK_MARKETING_API_TIMEOUT', 3)
```

File: lms/djangoapps/course_api/helpers.py [hunk 4]
```diff
@@ -53,12 +56,33 @@ def get_marketing_data(course_key, language):
...
+        response = requests.get(
+            url=url,
+            headers={'Accept-Language': language},
+            timeout=settings.EDRAAK_MARKETING_API_TIMEOUT
+        )
+    except requests.exceptions.Timeout:
+        raise APIException(
+            {
+                "status_code": status.HTTP_503_SERVICE_UNAVAILABLE,
+                "developerMessage": "Marketing courses API have timed out."
```
Comment (ghost) [inline lms/djangoapps/course_api/helpers.py]:
```suggestion
                "developer_message": "Marketing courses API have timed out."
```

File: lms/djangoapps/course_api/helpers.py [hunk 5]
```diff
@@ -53,12 +56,33 @@ def get_marketing_data(course_key, language):
...
+            },
+            code=status.HTTP_503_SERVICE_UNAVAILABLE,
+        )
+    except Exception:
+        log.exception('Something went wrong with the marketing courses API')
+        raise APIException(
+            {
+                "status_code": status.HTTP_500_INTERNAL_SERVER_ERROR,
+                "developerMessage": "Marketing courses didn't respond correctly."
```
Comment (ghost) [inline lms/djangoapps/course_api/helpers.py]:
```suggestion
                "developer_message": "Marketing courses didn't respond correctly."
```

File: lms/djangoapps/course_api/helpers.py [hunk 6]
```diff
@@ -53,12 +56,33 @@ def get_marketing_data(course_key, language):
...
+            },
+            code=status.HTTP_500_INTERNAL_SERVER_ERROR,
+        )
 
     if response.status_code != 200:
         log.warning('Could not fetch the marketing details from the API. course_key=[%s], status_code=[%s], url=[%s].',
                     course_key, response.status_code, url)
         return {}
+
```
Comment (ghost) [inline lms/djangoapps/course_api/helpers.py]:
```suggestion
```

File: lms/envs/aws.py [hunk 7]
```diff
@@ -708,6 +708,9 @@
 EDRAAK_AUTH_REDIRECT_ORIGINS_WHITELIST = ENV_TOKENS.get("EDRAAK_AUTH_REDIRECT_ORIGINS_WHITELIST", [])
 EDRAAK_AUTH_REDIRECT_REGX_ORIGINS = ENV_TOKENS.get("EDRAAK_AUTH_REDIRECT_REGX_ORIGINS", [])
 
+# API Calls Timeout
+EDRAAK_MARKETING_API_TIMEOUT = ENV_TOKENS.get("EDRAAK_MARKETING_API_TIMEOUT", 3)
```
Comment (ghost) [inline lms/envs/aws.py]:
```suggestion
EDRAAK_MARKETING_API_TIMEOUT = ENV_TOKENS.get('EDRAAK_MARKETING_API_TIMEOUT', 3)
```

File: cms/envs/aws.py (hunk 3 above)
Comment (ghost) [reply cms/envs/aws.py]: I'm not sure that we need this value in the CMS. Course API seems to be an LMS thing only.

**Review by ghost — APPROVED** [review]
Few minor changes and should be good to go. 

@devalih Please take a quick look before we merge this PR.

File: lms/djangoapps/course_api/helpers.py (hunk 4 above)
Comment (ghost) [reply lms/djangoapps/course_api/helpers.py]: Consistent with the other platform API. My fault for suggesting the camelcase style.

File: lms/djangoapps/course_api/helpers.py (hunk 5 above)
Comment (ghost) [reply lms/djangoapps/course_api/helpers.py]: Consistent with the other platform API. My fault for suggesting the camelcase style.

File: cms/envs/aws.py (hunk 3 above)
Comment (shadinaif) [reply cms/envs/aws.py]: Yes @OmarEdraak it is not needed in CMS. But tests will fail. There are two other options if you like it to be changed:
1- Add it to cms `test.py` . But makes no deference in my opinion
2- Mark the test as LMS only. Safe only if we are sure that we'll never use the link in CMS
or maybe a third option you have

Comment (shadinaif) [reply cms/envs/aws.py]: Oh! I've already added it to `test.py`. My bad, I'll remove it from from CMS. Resolving this as well

**Review by ghost — APPROVED** [review]

**Comment by ghost** [comment]
@shadinaif what's the status of this PR? Do we plan to merge it? // cc: @devalih

**Comment by shadinaif** [comment]
Closing PR for lack of interest to merge it. It's still there if needed again

**Pull request closed without merging**